src/tls_relation.py

```diff
@@ -244,73 +227,114 @@ def certificate_relation_available(  # type: ignore[no-untyped-def]
+            private_key=private_key_dict["key"].encode(),
+            private_key_password=private_key_dict["password"].encode(),
             subject=hostname,
+            sans_dns=[hostname],
         )
-        charm.certificates.request_certificate_renewal(
+        certificates.request_certificate_renewal(
             old_certificate_signing_request=old_csr.encode(),
             new_certificate_signing_request=new_csr,
         )
         self.update_relation_data_fields(
-            {f"csr-{hostname}": new_csr.decode()}, tls_certificates_relation, charm.app
+            {f"csr-{hostname}": new_csr.decode()}, tls_certificates_relation
         )
-        peer_relation = charm.model.get_relation(PEER_RELATION_NAME)
-        self.update_relation_data_fields(
-            {f"csr-{hostname}": new_csr.decode()}, peer_relation, charm.app
+        self.update_relation_data_fields({f"csr-{hostname}": new_csr.decode()}, peer_relation)
+
+    def get_decrypted_keys(self) -> Dict[str, str]:
+        """Return the list of decrypted private keys.
+
+        Returns:
+            A dictionary indexed by domain, containing the decrypted private keys.
+        """
+        decrypted_private_keys = {
+            str(hostname): self._get_decrypted_key(
+                (private_key := self._get_private_key(hostname))["key"], private_key["password"]
+            )
+            for hostname in self.keys
+            if hostname
+        }
+        return decrypted_private_keys
+
+    def _get_decrypted_key(self, private_key: str, password: str) -> str:
+        """Decrypted the provided private key using the provided password.
+
+        Args:
+            private_key: The encrypted private key.
+            password: The password to decrypt the private key.
+
+        Returns:
+            The decrypted private key.
+        """
+        decrypted_key = serialization.load_pem_private_key(
+            private_key.encode(), password=password.encode()
         )
+
+        # There are multiple representation PKCS8 is the default supported by nginx controller
+        return decrypted_key.private_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PrivateFormat.PKCS8,
+            encryption_algorithm=serialization.NoEncryption(),
+        ).decode()
+
+    def _get_private_key(self, hostname: str) -> Dict[str, str]:
+        """Return the private key and its password from either juju secrets or the relation data.
+
+        Args:
+            hostname: The hostname of the private key we want to fetch.
+
+        Returns:
+            The encrypted private key.
+        """
+        private_key_dict = {}
+        if JujuVersion.from_environ().has_secrets:
+            secret = self.charm_model.get_secret(label=f"private-key-{hostname}")
+            private_key_dict["key"] = secret.get_content()["key"]
+            private_key_dict["password"] = secret.get_content()["password"]
```

Comment:
If the key and the password are stored in the same place, why is the key encrypted?

Reply:
That's true, I didn't change the implementation here, might be interesting to store them in separate secrets. There's unfortunately no good way to separate the password from the key, similarly the key is dumped into k8s unencrypted. But I'd rather have from the beginning an encrypted key, than manipulating non encrypted certificates.